File: Part 8 - Deep Learning/Artificial Neural Networks (ANN)/Ann_Swats.py
#Theano and Tensorflow used to build deep learning networks from scratch. Mainly used in research work.
#Keras wraps the above two libraries. It is built on top of above libraries

#Importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Importing dataset
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

#Encode Categorical Data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_country = LabelEncoder()
X[:, 1] = labelencoder_X_country.fit_transform(X[:, 1])
labelencoder_X_gender = LabelEncoder()
X[:, 2] = labelencoder_X_gender.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]

#Splitting dataset into training and test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

#You need to apply feature scaling to ease all the calculations
#Feature Scaling - We use feature scaling so large variable set don't drown out smaller ones
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

#Building ANN
#Importing Keras libraries and packages
import keras
from keras.models import Sequential
from keras.layers import Dense

#Initializing ANN
classifier = Sequential()

#Adding input layer and an hidden layer
#Dense function - Activation function, Number of inputs in the input layer, no. of nodes to be added in hidden layer
#Intializing weights according to uniform distribution
#Rectifier activation func for hidden layer and sigmoid activation func for output layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu', input_dim = 11))

#Adding second hidden layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu'))

#Adding the output layer, binary outcomes have only one output layer, sigmoid function is the heart of activation function
#If we have more than 2 classes - 3 or more, then we use softmax function
classifier.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))

#Applying stochastic gradient to the neural network model - this process is called compiling
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

#Fitting ANN to the training set
classifier.fit(X_train, y_train, batch_size = 10, nb_epoch = 100)

#Predicting Test Set Results
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

#Making Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

File: Part 8 - Deep Learning/Artificial Neural Networks (ANN)/Ann_Swats.R
#Artificial Neural Network

# Importing the dataset
dataset = read.csv('Churn_Modelling.csv')
dataset = dataset[4:14]

# Encoding the categorical data as factors
dataset$Geography = as.numeric(factor(dataset$Geography,
                                      levels = c('France', 'Spain', 'Germany'),
                                      labels = c(1, 2, 3)))
dataset$Gender = as.numeric(factor(dataset$Gender,
                                  levels = c('Female', 'Male'),
                                  labels = c(1, 2)))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
#We need to apply feature scaling for training a neural network because it is computation intensive
training_set[-11] = scale(training_set[-11])
test_set[-11] = scale(test_set[-11])

# Fitting ANN to the Training set
#Neural - Deep learning models for regressors
#Nnet - Deep learning models with only one hidden layer
#DeepNet - Deep Learning models with many hidden layers
#H2O - Open source s/w platform, that allows connection to a instance of computer systems and allows to train the model efficiently
#      Gives lots of options to build a deep learning model
#     It contains parameter tuning arguments that helps to choose optimal numbers to build deep learning model
#install.packages('h2o')
library(h2o)
#Initialising the remote connection, nthreads - no. of cores in the system you want to connect to build deep learning model
# -1 - take all the cores/ optimize the no. of cores in the system
h2o.init(nthreads = -1)
classifier = h2o.deeplearning(y = 'Exited',
                              training_frame = as.h2o(training_set),
                              activation = 'Rectifier',
                              hidden = c(6, 6),
                              epochs = 100,
                              train_samples_per_iteration = -2)

# Predicting the Test set results
prob_pred = h2o.predict(classifier, newdata = as.h2o(test_set[-11]))
y_pred = (prob_pred > 0.5)
y_pred = as.vector(y_pred)

# Making the Confusion Matrix
cm = table(test_set[, 11], y_pred)

#Disconnect from the remote server
h2o.shutdown()